File: app/src/test/test/java/com/timeback/ui/viewmodel/DataManagementViewModelTest.java
package com.timeback.ui.viewmodel;

import com.timeback.ui.domain.model.*;
import com.timeback.ui.fake.DataManagementViewData;
import com.timeback.ui.fake.FakeFeatureGateway;
import com.timeback.ui.feature.datamanagement.DataManagementViewModel;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * DataManagementViewModel 단위 테스트.
 */
class DataManagementViewModelTest {

    private FakeFeatureGateway gateway;
    private DataManagementViewModel viewModel;

    @BeforeEach
    void setUp() {
        gateway = new FakeFeatureGateway();
    }

    @Test
    void loadState_shouldReturnContentWithDataManagementViewData() throws Exception {
        viewModel = new DataManagementViewModel(gateway);
        Thread.sleep(300);

        ScreenState state = viewModel.state.getValue();
        assertNotNull(state);
        assertInstanceOf(ScreenState.Content.class, state);

        Object data = ((ScreenState.Content) state).getData();
        assertInstanceOf(DataManagementViewData.class, data);
        DataManagementViewData dmvd = (DataManagementViewData) data;
        assertTrue(dmvd.isAnonymousIdReady());
        assertNotNull(dmvd.getRetentionSelection());
    }

    @Test
    void changeRetention_shouldReturnSuccess() throws Exception {
        viewModel = new DataManagementViewModel(gateway);
        Thread.sleep(300);

        viewModel.changeRetention(new RetentionSelection("7days"));
        Thread.sleep(300);

        ActionResult result = viewModel.actionResult.getValue();
        assertNotNull(result);
        assertInstanceOf(ActionResult.Success.class, result);
    }

    @Test
    void confirmDeletion_shouldSetDeletionComplete() throws Exception {
        viewModel = new DataManagementViewModel(gateway);
        Thread.sleep(300);

        viewModel.confirmDeletion();
        Thread.sleep(500);  // 삭제 확인 + 재로드

        ActionResult result = viewModel.actionResult.getValue();
        assertInstanceOf(ActionResult.Success.class, result);

        // 상태 재로드됨 - deletionComplete = true
        ScreenState state = viewModel.state.getValue();
        assertInstanceOf(ScreenState.Content.class, state);
        DataManagementViewData data = (DataManagementViewData) ((ScreenState.Content) state).getData();
        assertNotNull(data.getDeletionJob());
        assertTrue(data.getDeletionJob().isDeviceCompleted());
        assertTrue(data.getDeletionJob().isServerCompleted());
    }

    @Test
    void refreshDeletionStatus_shouldReturnUpdatedStatus() throws Exception {
        gateway.deletionComplete = true;
        viewModel = new DataManagementViewModel(gateway);
        Thread.sleep(300);

        viewModel.refreshDeletionStatus(new Identifier("job-1"));
        Thread.sleep(300);

        ScreenState state = viewModel.state.getValue();
        assertInstanceOf(ScreenState.Content.class, state);
        DataManagementViewData data = (DataManagementViewData) ((ScreenState.Content) state).getData();
        assertNotNull(data.getDeletionJob());
    }
}

File: src/test/java/com/timeback/ui/viewmodel/DataManagementViewModelTest.kt
package com.timeback.ui.viewmodel

import com.timeback.ui.domain.model.*
import com.timeback.ui.fake.DataManagementViewData
import com.timeback.ui.fake.FakeFeatureGateway
import com.timeback.ui.feature.datamanagement.DataManagementViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.*
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*

/**
 * UI-08 DataManagementViewModel 단위 테스트
 * 검증: CONTENT(backup pending), RETRYABLE_ERROR(offline), confirmFullDeletion flow, PARTIAL_FAILURE
 */
@OptIn(ExperimentalCoroutinesApi::class)
class DataManagementViewModelTest {

    private val testDispatcher = UnconfinedTestDispatcher()
    private lateinit var gateway: FakeFeatureGateway
    private lateinit var viewModel: DataManagementViewModel

    @BeforeEach
    fun setup() {
        Dispatchers.setMain(testDispatcher)
        gateway = FakeFeatureGateway()
    }

    @AfterEach
    fun tearDown() {
        Dispatchers.resetMain()
    }

    @Test
    fun `normal state shows CONTENT with backup pending`() = runTest {
        gateway.isOffline = false
        gateway.deletionComplete = false
        viewModel = DataManagementViewModel(gateway)
        advanceUntilIdle()

        val state = viewModel.state.value
        assertTrue(state is ScreenState.Content<*>)
        val data = (state as ScreenState.Content<*>).data as DataManagementViewData
        assertEquals(3, data.backupStatus.pendingCount)
    }

    @Test
    fun `offline shows RETRYABLE_ERROR`() = runTest {
        gateway.isOffline = true
        viewModel = DataManagementViewModel(gateway)
        advanceUntilIdle()

        val state = viewModel.state.value
        assertTrue(state is ScreenState.RetryableError)
    }

    @Test
    fun `requestFullDeletion sets confirmation pending`() = runTest {
        gateway.isOffline = false
        viewModel = DataManagementViewModel(gateway)
        advanceUntilIdle()

        viewModel.requestFullDeletion()
        advanceUntilIdle()

        assertTrue(viewModel.confirmationPending.value)
    }

    @Test
    fun `confirmFullDeletion completes and shows deletion done`() = runTest {
        gateway.isOffline = false
        viewModel = DataManagementViewModel(gateway)
        advanceUntilIdle()

        viewModel.requestFullDeletion()
        advanceUntilIdle()
        viewModel.confirmFullDeletion()
        advanceUntilIdle()

        assertFalse(viewModel.confirmationPending.value)
        val state = viewModel.state.value
        assertTrue(state is ScreenState.Content<*>)
        val data = (state as ScreenState.Content<*>).data as DataManagementViewData
        assertNotNull(data.deletionJob)
        assertTrue(data.deletionJob!!.deviceCompleted)
        assertTrue(data.deletionJob!!.serverCompleted)
    }

    @Test
    fun `changeRetention offline returns RETRYABLE_FAILURE`() = runTest {
        gateway.isOffline = true
        viewModel = DataManagementViewModel(gateway)
        advanceUntilIdle()

        // Reset to test action
        gateway.isOffline = true
        viewModel.changeRetention(RetentionSelection("30days"))
        advanceUntilIdle()

        val result = viewModel.actionResult.value
        assertTrue(result is ActionResult.RetryableFailure)
    }
}
